resources/hostname.go

```diff
@@ -0,0 +1,326 @@
+	callResult := bus.BusObject().Call(
+		"org.freedesktop.DBus.AddMatch", 0,
+		fmt.Sprintf("type='signal',path='%s',interface='org.freedesktop.DBus.Properties',member='PropertiesChanged'", hostname1Path))
+	if callResult.Err != nil {
+		return errwrap.Wrap(callResult.Err, "Failed to subscribe to DBus events for hostname1")
+	}
+
+	signals := make(chan *dbus.Signal, 10)
+	bus.Signal(signals)
+
+	var send = false // send event?
+	var dirty = false
+	var closed = false
+
+	// Ensure that we only close the channel when it wasn't closed already
+	defer func() {
+		if !closed {
+			close(signals)
+		}
+	}()
+
+	for {
+		obj.SetState(ResStateWatching) // reset
+		select {
+		case _, ok := <-signals:
+			cuid.SetConverged(false)
+
+			closed = !ok // Not OK means the channel has been closed
+			send = true
+			dirty = true
+
+		case event := <-obj.Events():
+			cuid.SetConverged(false)
+			// we avoid sending events on unpause
+			if exit, _ := obj.ReadEvent(&event); exit {
+				return nil // exit
+			}
+			send = true
+			dirty = true
+
+		case <-cuid.ConvergedTimer():
+			cuid.SetConverged(true) // converged!
+			continue
+
+		case <-Startup(startup):
+			cuid.SetConverged(false)
+			send = true
+		}
+
+		// do all our event sending all together to avoid duplicate msgs
+		if send {
+			startup = true // startup finished
+			send = false
+			if dirty {
+				dirty = false
+				obj.isStateOK = false // something made state dirty
+			}
+			// only do this on certain types of events
+			if exit, err := obj.DoSend(processChan, ""); exit || err != nil {
+				return err // we exit or bubble up a NACK...
+			}
+		}
+		if closed {
+			return errwrap.New("DBUS channel has been closed")
+		}
+	}
+}
+
+func updateHostnameProperty(object dbus.BusObject, expectedValue, property, setterName string, apply bool) (bool, error) {
+	propertyObject, err := object.GetProperty("org.freedesktop.hostname1." + property)
+	if err != nil {
+		return false, errwrap.Wrapf(err, "failed to get org.freedesktop.hostname1.%s", property)
+	}
+	if propertyObject.Value() == nil {
+		return false, errwrap.Errorf("Unexpected nil value received when reading property %s", property)
+	}
+
+	propertyValue, ok := propertyObject.Value().(string)
+	if !ok {
+		return false, fmt.Errorf("Received unexpected type as %s value, expected string got '%T'", property, propertyValue)
+	}
+
+	// Expected value and actual value match, nothing todo => true
+	if propertyValue == expectedValue {
+		return true, nil
+	}
+
+	// Values don't match but shouldn't be changed either => false
+	if !apply {
+		return false, nil
+	}
+
+	// Attempting to apply the changes
+	log.Printf("Changing %s: %s => %s", property, propertyValue, expectedValue)
+	if err := object.Call("org.freedesktop.hostname1."+setterName, 0, expectedValue, false).Err; err != nil {
+		return false, errwrap.Wrapf(err, "failed to call org.freedesktop.hostname1.%s", setterName)
+	}
+
+	// All good changes should now be applied again
+	return true, nil
```

Comment:
Since we did make changes (successfully) this should be `return false, nil`. This is assuming you're going for the same pattern as the CheckApply signature.